homer0/projext
# fix: improve Babel settings for polyfills

### What does this PR do?

> On #77, I didn't understand that the env preset could take care of the polyfills when configured with the version of `core-js`.

I moved back the functionality of the `babel.polyfill` setting to the core package: When the setting is `true`, projext will now add the `corejs` and `useBuiltIns` settings the env preset needs in order to use `core-js` and `regenerator-runtime` directly. There's no more need for the build engines to take care of that.

While writing that, I realized that the env preset has grown to the point that most of Babel's behavior is controlled from there, and changing its settings wasn't that easy with the `babel.overwrites` setting.... so I added a new target setting: `babel.env`.

The new setting, for both browser and Node targets, is used as a base for the preset configuration projext creates, so you can make any change without having to declare the plugins array with `babel.overwrites`.

This is only for browser targets: projext will check if the `env` setting has a `targets.browsers` list, and if it already has rules for some browsers, it will only add rules for the ones that are not present.

### How should it be tested manually?

The true value of this is that projext will no longer include the entire `core-js` package, but in order to test this, you have to do some "hacking":

First, run `projext analyze` on your target to see how much space is taking `core-js`, then... 

1. Install this version of projext and any of the build engines.
2. Delete the projext version installed inside the build engine `node_modules` (since this is not installed from a "proper version", `npm`/`yarn`, may install the one required by the build engine inside its directory: `node_modules/projext-plugin-[build-engine]/node_modules/projext`).
3. Open the `polyfill.js` file on the root of the build engine directory and remove all its contents.

Now, run `analyze` again and `core-js` should be "smaller" than before, as now it's only including what it's actually needed.

Also, don't forget to...

```bash
npm test
# or
yarn test
```
